test/Fencer/Server/Test.hs

```diff
@@ -87,32 +87,53 @@ test_serverResponseNoRules =
 -- gRPC server
 ----------------------------------------------------------------------------
 
+-- | A type combining a logger, a handle for the logging file, a
+-- thread id and an application state.
+type LogIdSt = (Logger.Logger, Handle, ThreadId, AppState)
+
 -- | Start Fencer on the default port.
-createServer :: IO (Logger.Logger, ThreadId)
+createServer :: IO (Logger.Logger, Handle, ThreadId)
 createServer = do
-  (logger, threadId, _) <- createServerAppState
-  pure (logger, threadId)
+  (logger, logHandle, threadId, _) <- createServerAppState
+  pure (logger, logHandle, threadId)
 
 -- | Start Fencer on the default port.
-createServerAppState :: IO (Logger.Logger, ThreadId, AppState)
+createServerAppState :: IO LogIdSt
 createServerAppState = do
   -- TODO: not the best approach. Ideally we should use e.g.
   -- https://hackage.haskell.org/package/tasty-hunit/docs/Test-Tasty-HUnit.html#v:testCaseSteps
   -- but we can't convince @tinylog@ to use the provided step function.
-  logger <- Logger.create (Logger.Path "/dev/null")
+
+  tmpDir <- Temp.getCanonicalTemporaryDirectory
+  -- This opens a temporary file in the ReadWrite mode
+  (loggerPath, logHandle) <- Temp.openTempFile tmpDir "fencer-server.log"
+  -- The handle has to be closed. Otherwise trying to create a logger
+  -- would fail due to a file lock.
+  hClose logHandle
```

Comment:
This is kind of weird. Would it be better to use `emptyTempFile` here?

Reply:
I gave it a try, but then I couldn't find a way to make it all work together. With it I don't get a handle and how would I close it later on? My understanding is that the return value `IO FilePath` with the documentation "Path to the (written and closed) file" means that I have to open the file after calling `emptyTempFile`.

Reply:
> With it I don't get a handle and how would I close it later on?

I think, the handle you're closing there is the one that you've already closed, i.e. it's a redundant operation. What if you just remove `logHandle` from the state, would it change anything? Quoting the docs:

> A handle is open when first allocated. Once it is closed it can no longer be used for either input or output, though an implementation cannot re-use its storage while references remain to it.

Reply:
I admit there might be something fishy here, but I am not sure what. Before I put that `hClose` there I would get a runtime error in the line trying to create a logger, which would report a file is locked. Per my comment in the code, closing it just the line before removes that error.

Reply:
Missed your reply somehow. Yeah, that makes perfect sense to me, but that's what should be fixed by using `emptyTempFile`. And what I'm asking is whether you can remove `logHandle` from the state entirely and not break anything.

Reply:
Maybe there's a better way, but I don't see it. The thing is that in tests I want to log to a temporary file and so far I haven't found a better way than what I have in the code. Probably the main limiting factor is using `withServer :: (IO Server -> TestTree) -> TestTree`; I can't figure out how to use it in conjunction with `emptyTempFile`.

For this to work I think we would have to use a more advanced logger (we're using `tinylog`) that would allow us to simply log nothing during testing.